src/plugins/intel_gpu/src/graph/graph_optimizer/reshape_transfer.cpp

```diff
@@ -0,0 +1,141 @@
+                return is_suitable_parent(&input);
+            } else if (input.is_type<activation>()) {
+                return is_suitable_parent(&input);
+            }
+            return false;
+        }
+        return false;
+    };
+
+    auto update_order = [](std::vector<uint16_t> original_order, cldnn::program_node* reshape) {
+        if (!reshape)
+            return original_order;
+        // Example. For this sequence, there is Reshape node which merges 2 consecutive dims into one
+        // order must be updated like permute is done before reshape
+        // [1,3,4,6] -> Reshape[1,3,24,1]-> permute(0,2,1) -> [1,24,3,1]
+        // updated order must be (0,2,3,1):
+        // dim with index=2 is split into 2 parts: 2 and 3
+        const auto& reshape_in_shape = reshape->get_input_layout().get_dims();
+        const auto& reshape_out_shape = reshape->get_output_layout().get_dims();
+        auto transformed_order = original_order;
+        ov::Shape new_shape(transformed_order.size());
+        const uint16_t merge_dim_idx = [&]() {
+            for (uint16_t i = 0; i < reshape_in_shape.size(); ++i) {
+                if (reshape_in_shape[i] != reshape_out_shape[i])
+                    return i;
+            }
+            OPENVINO_THROW("same input/output for reshape node");
+        }();
+        auto insertIt = transformed_order.end();
+        for (auto it = transformed_order.begin(); it != transformed_order.end(); ++it) {
+            auto& elem = *it;
+            if (elem > merge_dim_idx) {
+                elem++;
+            } else if (elem == merge_dim_idx) {
+                insertIt = it + 1;
+            }
+        }
+        transformed_order.insert(insertIt, merge_dim_idx + 1);
+        // remove invalid orders
+        if (transformed_order.size() > reshape_out_shape.size()) {
+            transformed_order.erase(
+                std::remove_if(transformed_order.begin(), transformed_order.end(), [&](uint16_t& order) {
+                    return order >= reshape_out_shape.size();
+                }));
+        }
+        return transformed_order;
+    };
+
+    auto itr = p.get_processing_order().begin();
+    while (itr != p.get_processing_order().end()) {
+        auto& node = *itr++;
+        if (!node->is_type<permute>())
+            continue;
+        auto& transpose_node = node->as<permute>();
+        if (!is_suitable_permute(&transpose_node))
+            continue;
+        auto& child_node = transpose_node;
+        auto parent_node = child_node.get_dependency_with_port(0).first;
+        cldnn::program_node* inter_node;
+        if (parent_node->is_type<reshape>()) {
+            inter_node = parent_node;
+            if (!is_suitable_reshape(inter_node)) {
+                continue;
+            }
+            parent_node = inter_node->get_dependency_with_port(0).first;
+        } else {
+            continue;
+        }
+
+        if (!is_suitable_parent(parent_node)) {
```

Comment:
Will this recursive function lead to long latency in some case?

Reply:
opt the recursive for more strict pattern check